File: baekjoon/silver/1_1679_숫자놀이/숫자놀이.js
// 실버 1 : https://www.acmicpc.net/problem/1679

const fs = require("fs");
const file = process.platform === "linux" ? "/dev/stdin" : "./text.txt";
const input = fs.readFileSync(file).toString().trim().split("\n");

const n = Number(input[0]);
let nums = input[1].split(" ").map(Number);

nums.sort((a, b) => a - b);

const max_cnt = Number(input[2]);
const max_num = nums[nums.length - 1] * max_cnt;
let dp = Array(max_num + 1).fill(0);

for (let i = 1; i < max_num; i++) {
  if (i - nums[0] >= 0) {
    dp[i] = dp[i - nums[0]] + 1;
  }

  for (let j = 1; j < nums.length; j++) {
    if (i - nums[j] >= 0) {
      if (Math.min(dp[i], dp[i - nums[j]] + 1) <= max_cnt) {
        dp[i] = Math.min(dp[i], dp[i - nums[j]] + 1);
      }
    }
  }
}

for (let i = 0; i < dp.length; i++) {
  if (dp[i] > max_cnt) {
    if (i % 2 === 0) {
      console.log("holsoon win at", i);
      return;
    } else {
      console.log("jjaksoon win at", i);
      return;
    }
  }
}

File: baekjoon/silver/1_1679_숫자놀이/숫자놀이.py
# 실버 1 : https://www.acmicpc.net/problem/1679

n = int(input())
nums = list(map(int, input().split()))
nums.sort()
max_cnt = int(input())

max_nums = max(nums) * max_cnt
dp = [0]*(max_nums+1)

for i in range(1, max_nums+1):
  if i-nums[0] >= 0:
    dp[i] = dp[i-nums[0]] + 1
  for j in range(1, len(nums)):
    if i - nums[j] >= 0:
      if min(dp[i], dp[i - nums[j]]+1) <= max_cnt:
        dp[i] = min(dp[i], dp[i - nums[j]]+1)

for i, v in enumerate(dp):
  if v > max_cnt:
    if i % 2 == 0:
      print("holsoon win at", i)
      break
    else:
      print("jjaksoon win at", i)
      break
